lib/Driver/ToolChains.cpp

```diff
@@ -1101,33 +1101,49 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
 
   ArgStringList Arguments;
 
+  // Explicitly pass the target to the linker
+  Arguments.push_back(context.Args.MakeArgString("--target=" + getTriple().str()));
+
   switch (job.getKind()) {
   case LinkKind::None:
     llvm_unreachable("invalid link kind");
   case LinkKind::Executable:
-    // Default case, nothing extra needed
     break;
   case LinkKind::DynamicLibrary:
     Arguments.push_back("-shared");
-    if (getTriple().getOS() == llvm::Triple::Linux) {
-      if (getTriple().getSubArch() == llvm::Triple::SubArchType::ARMSubArch_v7 ||
-          getTriple().getSubArch() == llvm::Triple::SubArchType::ARMSubArch_v6) {
-        Arguments.push_back("-Wl,-Bsymbolic");
-      }
-    }
     break;
   }
 
-  addPrimaryInputsOfType(Arguments, context.Inputs, types::TY_Object);
-  addInputsOfType(Arguments, context.InputActions, types::TY_Object);
+  if (getTriple().getObjectFormat() == llvm::Triple::ELF) {
+    Arguments.push_back("-Wl,-Bsymbolic");
+  }
 
-  context.Args.AddAllArgs(Arguments, options::OPT_Xlinker);
-  context.Args.AddAllArgs(Arguments, options::OPT_linker_option_Group);
-  context.Args.AddAllArgs(Arguments, options::OPT_F);
+  // Select the linker to use
+  llvm::SmallString<128> Linker;
 
-  if (!context.OI.SDKPath.empty()) {
-    Arguments.push_back("--sysroot");
-    Arguments.push_back(context.Args.MakeArgString(context.OI.SDKPath));
+  switch (getTriple().getArch()) {
+  default:
+    break;
+  case llvm::Triple::arm:
+  case llvm::Triple::armeb:
+  case llvm::Triple::thumb:
+  case llvm::Triple::thumbeb:
+    // BFD linker has issues wrt relocation of the protocol conformance
+    // section on these targets, it also generates COPY relocations for
+    // final executables, as such, unless specified, we default to gold
+    // linker.
+    if (!context.Args.hasArg(options::OPT_use_ld)) {
+      Linker = "gold";
+    }
+    break;
+  }
+
+  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld)) {
+    Linker = A->getValue();
+  }
```

Comment:
I would move the switch into an else branch here. Also, since all of the strings will outlast this function, I suggest using StringRef instead of SmallString, to avoid an extra copy.